precompiles/Modexp.yul

```diff
@@ -642,6 +643,70 @@ object "ModExp" {
+                //     if modulus = 1 then
+                //         return 0
+                //     Assert :: (modulus - 1) * (modulus - 1) does not overflow base
+                //     result := 1
+                //     base := base mod modulus
+                //     while exponent > 0 do
+                //         if (exponent mod 2 == 1) then
+                //             result := (result * base) mod modulus
+                //         exponent := exponent >> 1
+                //         base := (base * base) mod modulus
+                //     return result
+                
+                // PSEUDOCODE: `if modulus = 1 then return 0`.
+                // We are using the precondition that `result[] == [0, ..., 0]`.
+                if big_uint_is_not_one(n_limbs, ptr) {
+
+                    // Assert :: (modulus - 1) * (modulus - 1) does not overflow base
+                    // We are certain that this is true because our precondition requires the most significant half of exponent to be zeros.
+
+                    // PSEUDOCODE: `result := 1`
+                    // Again, we are using the precondition that `result[] == 0`
+                    bigUIntInPlaceOrWith1(result_ptr, n_limbs)
+
+                    // PSEUDOCODE: `base := base mod modulus`
+                    // FIXME: Is ok to mutate the base[] we were given? Shall we use a temporal buffer?
+                    big_uint_mod_inplace(n_limbs, base_ptr, modulus_ptr) // FIXME: not yet implemented.
+
+                    // PSEUDOCODE: `while exponent > 0 do`
+                    // FIXME: Is ok to mutate the exponent[] we were given? Shall we use a temporal buffer?
+                    for { } big_uint_is_not_zero(exponent) { } {
+
+                        // PSEUDOCODE: `if (exponent mod 2 == 1) then`
+                        if big_uint_mod_two(n_limbs, exponent_ptr) {
+
+                            // PSEUDOCODE: `result := (result * base) mod modulus`
+                            // Since result[] is our return value, we are allowed to mutate it.
+                            big_uint_mul_mod_inplace(n_limbs, result_ptr, base_ptr, modulus_ptr) // Fixme: not yet implemented. 
+
+                        }
+
+                        // PSEUDOCODE: `exponent := exponent >> 1`
+                        // FIXME: Is ok to mutate the exponent[] we were given? Shall we use a temporal buffer?
+                        bigUIntOneShiftRight(exponent_ptr, n_limbs)
+                        
+                        // PSEUDOCODE: `base := (base * base) mod modulus`
+                        bigUIntOneShiftLeft(base_ptr, n_limbs) // base := base * base == base**2 == base << 1
+
+                        // Fixme: not yet implemented. 
+                        // FIXME: Is ok to mutate the base[] we were given? Shall we use a temporal buffer?
+                        big_uint_mod_inplace(n_limbs, base_ptr, modulus_ptr) // base:= base % modulus
```

Comment:
Same as before:

We shall discuss if it is **possible** and **acceptable** to mutate `base[]`, which was passed as a parameter.

This depends on how will this function will be used.

- This is to prevent us from requiring a temporary buffer. is it possible? to perform mod in place without one?
  - This helper function will need to call _divmod_ which already requires some temporary buffers.
    - Maybe this memory optimization i'm proposing does not make sense.